### PR TITLE
Updates for getting attribute info - #42

### PR DESCRIPTION
**Pull request by mjb3279**
- Updated comment for `get_attribute_info`
- Added `get_attribute_info_map`
- Added `get_single_attribute_info`
- Added `Hash` derivation for `AttributeType`

Signed-off-by: Michael J Boquard <michael.j.boquard@gmail.com>

### CODE REVIEW

**Review by hug-dev — COMMENTED** [review]
Thanks! A few more ideas...

File: cryptoki/src/functions/object_management.rs [hunk 1]
```diff
@@ -132,6 +193,86 @@ impl<'a> Session<'a> {
...
+    ///
+    /// The return code may be [Rv::Ok] or an `Rv::Error` containing either
+    /// [RvError::AttributeTypeInvalid] (if the attribute is not a valid attribute for the object),
+    /// or a [RvError::AttributeSensitive] (if the attribute is a sensitive attribute for an
+    /// object marked as sensitive, e.g. the private exponent of an RSA key)
+    pub fn get_single_attribute_info(
+        &self,
+        object: ObjectHandle,
+        attribute_type: AttributeType,
+    ) -> Result<(Rv, AttributeInfo)> {
```
[the remark this answers is not in the document]
Comment (mjb3279) [reply cryptoki/src/functions/object_management.rs:246]: What if we changed `get_attribute_info` to instead call `C_GetAttributeInfo` for each `AttributeType`, individually and update `AttributeInfo` to include the `TypeInvalid` and `Sensitive` variants.

It wouldn't really provide a 1-to-1 API wrapper, as you would be calling `C_GetAttributeInfo` `n` times with `n` being the number of `AttributeType` entries in the provided slice.

The other concern with the proposed solution is whether or not that breaks any application currently using `rust-cryptoki`

[the remark this answers is not in the document]
Comment (mjb3279) [reply cryptoki/src/functions/object_management.rs:246]: Alternatively we could have a `get_attribute_info_detailed` and return a different enum containing all the possible variants.  However I feel like that just adds 'clutter' to the code.

[the remark this answers is not in the document]
Comment (hug-dev) [reply cryptoki/src/functions/object_management.rs:246]: > What if we changed get_attribute_info to instead call C_GetAttributeInfo for each AttributeType, individually and update AttributeInfo to include the TypeInvalid and Sensitive variants.

Oh yes that's a pretty good idea!! I think it's fine not to provide a 1-1 mapping if in that case it makes things easier to use in my opinion (and overcome limitations of the API).

> The other concern with the proposed solution is whether or not that breaks any application currently using rust-cryptoki

I think that's fine at this point: the crate is young and we are under the first stable (still `0.x.y` versions).

If we go this way and break the API anyway, I think it's fine to just have one `get_attribute_info` returning the `HashMap` with the updated `AttributeType` (and delete the two other variants).

[the remark this answers is not in the document]
Comment (mjb3279) [reply cryptoki/src/functions/object_management.rs:246]: > overcome limitations of the API

If that's the goal, we have a _LOT_ of work to do 😉 

I'll get working on the proposed change.

File: cryptoki/src/functions/object_management.rs [hunk 2]
```diff
@@ -97,39 +98,127 @@ impl<'a> Session<'a> {
...
+    ///
+    /// # Returns
+    ///
+    /// This function will return a HashMap of [AttributeType] and [AttributeInfo] enums that will
+    /// either contain the size of the requested attribute or [AttributeInfo::Unavailable] if the
+    /// attribute is not available to be read from `object`.
+    ///
+    /// _Note: An attribute returning [AttributeInfo::Unavailable] may mean that the attribute is
+    /// either sensitive or not a valid type for `object`.  This function does not distinguish
+    /// between those two errors._
```
Comment (hug-dev) [inline cryptoki/src/functions/object_management.rs:209]: I guess now this note can be removed?

**Review by hug-dev — APPROVED** [review]
Thanks 👌

**Review by ionut-arm — APPROVED** [review]
Looks good, thanks!

**Pull request merged**